File: pkg/domain/model/runtime_manifest.go
package model

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
)

// RuntimeManifestFormatV1 is the format identifier of a runtime manifest
// artifact: a dbt partial-parse msgpack file produced by compiling a service's
// dbt project. It is stamped on every descriptor so a consumer can reject an
// artifact whose layout it does not understand.
const RuntimeManifestFormatV1 = "dbt-partial-parse-msgpack-v1"

// RuntimeManifestRef points at the runtime manifest a node should execute
// against. It is embedded in the events services exchange, so all four fields
// are set together or none are: a partial reference is a contract violation,
// not a degraded mode.
type RuntimeManifestRef struct {
	// RuntimeManifestURI is the s3:// location of the artifact.
	RuntimeManifestURI string `json:"runtime_manifest_uri,omitempty"`
	// RuntimeManifestSHA256 is the artifact's content digest, used both to
	// verify the download and to name the worker pool that serves it.
	RuntimeManifestSHA256 string `json:"runtime_manifest_sha256,omitempty"`
	// RuntimeManifestDBTVersion is the dbt-core version that produced the
	// artifact. A partial parse is only loadable by the version that wrote it.
	RuntimeManifestDBTVersion string `json:"runtime_manifest_dbt_version,omitempty"`
	// RuntimeManifestParseContextSHA256 digests the parse context (command
	// dialect plus target/environment) the artifact was produced under. A
	// consumer whose own context hashes differently must not reuse it.
	RuntimeManifestParseContextSHA256 string `json:"runtime_manifest_parse_context_sha256,omitempty"`
}

// Complete reports whether every field of the reference is populated.
func (r RuntimeManifestRef) Complete() bool {
	return r.RuntimeManifestURI != "" &&
		r.RuntimeManifestSHA256 != "" &&
		r.RuntimeManifestDBTVersion != "" &&
		r.RuntimeManifestParseContextSHA256 != ""
}

// Validate accepts the zero reference (meaning "no runtime manifest") and any
// complete, well-formed one. Anything in between is rejected so a half-filled
// reference cannot silently reach an executor.
func (r RuntimeManifestRef) Validate() error {
	if r == (RuntimeManifestRef{}) {
		return nil
	}
	if !r.Complete() {
		return fmt.Errorf("runtime manifest reference is partial")
	}
	if !strings.HasPrefix(r.RuntimeManifestURI, "s3://") {
		return fmt.Errorf("runtime_manifest_uri must be s3://")
	}
	digests := []struct {
		name  string
		value string
	}{
		{"runtime_manifest_sha256", r.RuntimeManifestSHA256},
		{"runtime_manifest_parse_context_sha256", r.RuntimeManifestParseContextSHA256},
	}
	for _, d := range digests {
		if err := validateSHA256Hex(d.name, d.value); err != nil {
			return err
		}
	}
	return nil
}

// validateSHA256Hex requires exactly 64 lowercase hex characters. Digests are
// compared verbatim across services, so an uppercase or truncated spelling of
// the same value would read as a different artifact.
func validateSHA256Hex(name, value string) error {
	if len(value) != 64 || strings.ToLower(value) != value {
		return fmt.Errorf("%s must be lowercase SHA-256 hex", name)
	}
	if _, err := hex.DecodeString(value); err != nil {
		return fmt.Errorf("%s must be lowercase SHA-256 hex: %w", name, err)
	}
	return nil
}

// RuntimeManifestDescriptor is the full record of a published runtime manifest
// artifact: what it is, which release and image produced it, and the context it
// was parsed under. Producers publish the descriptor; consumers carry the
// narrower RuntimeManifestRef.
type RuntimeManifestDescriptor struct {
	Format             string `json:"format"`
	ServiceName        string `json:"service_name"`
	ReleaseID          string `json:"release_id"`
	ImageTag           string `json:"image_tag"`
	ArtifactURI        string `json:"artifact_uri"`
	SHA256             string `json:"sha256"`
	DBTCoreVersion     string `json:"dbt_core_version"`
	AdapterType        string `json:"adapter_type"`
	ParseContextSHA256 string `json:"parse_context_sha256"`
}

// Ref narrows the descriptor to the reference that travels on events.
func (d RuntimeManifestDescriptor) Ref() RuntimeManifestRef {
	return RuntimeManifestRef{
		RuntimeManifestURI:                d.ArtifactURI,
		RuntimeManifestSHA256:             d.SHA256,
		RuntimeManifestDBTVersion:         d.DBTCoreVersion,
		RuntimeManifestParseContextSHA256: d.ParseContextSHA256,
	}
}

// WorkerPoolKey names the pool of reusable workers that can serve a node: the
// workers sharing a service, a team image, and a runtime manifest are
// interchangeable, and any difference in those three needs a distinct pool. The
// NUL separator keeps the fields unambiguous, so no two different triples can
// concatenate onto the same key.
func WorkerPoolKey(serviceName, imageTag, manifestSHA string) string {
	sum := sha256.Sum256([]byte(serviceName + "\x00" + imageTag + "\x00" + manifestSHA))
	return hex.EncodeToString(sum[:])
}

File: manifest-controller/service/runtime_manifest.py
"""Validation of the runtime manifest descriptor that travels beside a manifest.

A descriptor is the only thing read before a pre-built dbt partial parse is
accepted into a release, so every field is checked here: a descriptor that is
merely plausible would let a worker hydrate an artifact belonging to another
service, another release, or another dbt version.
"""
import re

from domain.exceptions import MalformedRuntimeManifestError
from domain.model import RuntimeManifestRef

# The layout of the artifact a descriptor describes: a dbt partial-parse msgpack
# file. A descriptor stamped with anything else describes contents this service
# cannot reason about, so it is rejected rather than guessed at.
FORMAT_V1 = "dbt-partial-parse-msgpack-v1"

# The only warehouse adapter the platform runs; an artifact parsed against a
# different adapter is not interchangeable with one parsed against postgres.
ADAPTER_TYPE = "postgres"

REQUIRED_FIELDS: tuple[str, ...] = (
    "format",
    "service_name",
    "release_id",
    "image_tag",
    "artifact_uri",
    "sha256",
    "dbt_core_version",
    "adapter_type",
    "parse_context_sha256",
)

_DIGEST_FIELDS: tuple[str, ...] = ("sha256", "parse_context_sha256")

# Digests are compared verbatim across services, so an uppercase or truncated
# spelling of the same value would read as a different artifact.
_SHA256_HEX = re.compile(r"^[0-9a-f]{64}$")


def parse_descriptor(
    descriptor: dict,
    expected_service: str,
    expected_artifact_uri: str,
) -> RuntimeManifestRef:
    """Validate a descriptor and narrow it to the ref that travels on events.

    Raises MalformedRuntimeManifestError if the descriptor is not a v1 descriptor
    for expected_service describing the artifact at expected_artifact_uri.
    """
    if not isinstance(descriptor, dict):
        raise MalformedRuntimeManifestError(
            f"runtime manifest descriptor: expected a JSON object, got {type(descriptor).__name__}"
        )

    for field in REQUIRED_FIELDS:
        if field not in descriptor:
            raise MalformedRuntimeManifestError(
                f"runtime manifest descriptor: {field} is missing"
            )
        if not isinstance(descriptor[field], str):
            raise MalformedRuntimeManifestError(
                f"runtime manifest descriptor: {field} must be a string"
            )
        if not descriptor[field]:
            raise MalformedRuntimeManifestError(
                f"runtime manifest descriptor: {field} is empty"
            )

    if descriptor["format"] != FORMAT_V1:
        raise MalformedRuntimeManifestError(
            f"runtime manifest descriptor: unsupported format {descriptor['format']!r},"
            f" expected {FORMAT_V1!r}"
        )

    if descriptor["service_name"] != expected_service:
        raise MalformedRuntimeManifestError(
            f"runtime manifest descriptor: service_name {descriptor['service_name']!r}"
            f" does not match declared service {expected_service!r}"
        )

    if descriptor["artifact_uri"] != expected_artifact_uri:
        raise MalformedRuntimeManifestError(
            f"runtime manifest descriptor: artifact_uri {descriptor['artifact_uri']!r}"
            f" is not the sibling artifact {expected_artifact_uri!r}"
        )

    for field in _DIGEST_FIELDS:
        if not _SHA256_HEX.fullmatch(descriptor[field]):
            raise MalformedRuntimeManifestError(
                f"runtime manifest descriptor: {field} must be lowercase SHA-256 hex"
            )

    if descriptor["adapter_type"] != ADAPTER_TYPE:
        raise MalformedRuntimeManifestError(
            f"runtime manifest descriptor: adapter_type {descriptor['adapter_type']!r}"
            f" is not {ADAPTER_TYPE!r}"
        )

    return RuntimeManifestRef(
        uri=descriptor["artifact_uri"],
        sha256=descriptor["sha256"],
        dbt_version=descriptor["dbt_core_version"],
        parse_context_sha256=descriptor["parse_context_sha256"],
    )
